docs/aot-jit-switch.md

```diff
@@ -0,0 +1,184 @@
+_Prop3_.
+ 
+To understand this pair of rules it is important to remember that a specific version of
+ErgoTree (in this case v1) assumes the fixed semantics of all operations. This however
+doesn't restrict the interpreter implementations and we use this fact
+to switch from `R4.0-AOT-verify` to `R5.0-JIT-verify` relying on their equivalence
+property _Prop 3_.
+
+However, for backward compatibility with Apps we DON'T NEED equivalence of costing, hence 
+exact cost estimation is not necessary. For this reason we have the relaxed condition in 
+_Prop4_, which means that any ScriptV1 admitted by `R4.0-AOT-cost` will also be admitted by
+`R5.0-JIT-cost`. For this reason, the v4.0 based application interacting with v5.0 node
+will not notice the difference.
+
+#### Rules 11 and 12
+After SF is activated v4.0 node cannot verify transactions containing v2 scripts, as
+a result the v4.0 node cannot include such transactions in new `candidate` blocks. Thus it
+performs `skip-pool-tx` action essentially using only those mempool transactions which it
+can handle. Majority of network nodes (v5.0) will do `R5.0-JIT-verify` validation of v2
+scripts in `candidate` blocks.
+
+#### Rules 13 and 14
+After SF is activated v4.0 node `skip-accept` verification of `mined` blocks by relying on
+the majority of v5.0, essentially not performing verification even of v1 scripts. The
+majority of nodes uses new JIT based implementation of ErgoTree interpreter
+`R5.0-JIT-verify` procedure for costing and verification, have consensus about blocks and
+v4.0 nodes just accept all mined blocks created elsewhere.
+
+#### Rules 15 and 16
+After SF is activated v4.0 node `skip-accept` of `mined` blocks by relying on the majority
+of v5.0 nodes since it cannot handle v2 scripts. In the same context, the majority of
+nodes uses the new JIT based implementation of ErgoTree interpreter `R5.0-JIT-verify`
+procedure for costing and verification.
+
+
+### Equivalence Properties of Validation Actions
+
+In order to guarantee network consensus in the presence of both v4.0 and v5.0 nodes
+the implementation of `R4.0-AOT-verify`, `R5.0-AOT-verify`, `R4.0-AOT-cost`,
+`R5.0-AOT-cost`, `R5.0-JIT-verify` should satisfy the following properties.
+
+_Prop 1._ AOT-verify is preserved: `forall s:ScriptV1, R4.0-AOT-verify(s) == R5.0-AOT-verify(s)` 
+
+_Prop 2._ AOT-cost is preserved: `forall s:ScriptV1, R4.0-AOT-cost(s) == R5.0-AOT-cost(s)`
+
+_Prop 3._ JIT-verify can replace AOT-verify: `forall s:ScriptV1, R5.0-JIT-verify(s) == R4.0-AOT-verify(s)`
+
+_Prop 4._ JIT-cost is bound by AOT-cost: `forall s:ScriptV1, R5.0-JIT-cost(s) <= R4.0-AOT-cost(s)`
+
+_Prop 5._ ScriptV2 is rejected before SF is active:  
+`forall s:ScriptV2, if not SF is active => R4.0-verify(s) == R5.0-verify(s) == Reject`
+
+### Other Notes
+
+- Since we are going to fix some bugs, the behaviour of v1 and v2 scripts in general not
+required to be precisely equivalent. So while old apps are supported unchanged, new apps
```

Comment:
Sounds contradicting to Prop. 3.

Reply:
Added the following clarification:

This is because `R5.0-JIT-verify` supports both v1
and v2 scripts so that `R5.0-JIT-verify(Script_v1) == R4.0-AOT-verify(Script_v1)` due to
[Prop 1](#equivalence-properties-of-validation-actions) and `R5.0-JIT-verify(Script_v2)`
may be completely different language. Of cause, we don't want it to be _completely_
different, in particular to ease migration.